### PR TITLE
edk2: backport OpenSSL 1.1.1t to the tree

### PR DESCRIPTION
**Pull request by RaitoBezarius**
###### Description of changes

Original bug: https://bugzilla.tianocore.org/show_bug.cgi?id=4342

We decide to backport this way because we do not have a lot of choices here, upgrading will break 23.05.

Prior art: https://github.com/NixOS/nixpkgs/pull/242473#issuecomment-1646247334.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by RaitoBezarius** [comment]
Please test well this PR.

**Comment by LunNova** [comment]
Result of `nixpkgs-review pr 244727` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>OVMF</li>
    <li>OVMF.fd</li>
    <li>OVMFFull</li>
    <li>OVMFFull.fd</li>
    <li>edk2</li>
    <li>edk2-uefi-shell</li>
    <li>multipass</li>
    <li>quickemu</li>
  </ul>
</details>

**Comment by LunNova** [comment]
`nix-build -A nixosTests.boot.uefiCdrom` passes on this branch

**Comment by LunNova** [comment]
@ofborg build nixosTests.boot.uefiCdrom

**Comment by LunNova** [comment]
Should we add that to passthru.tests for edk2?

**Review by LunNova — APPROVED** [review]
Test that broke with the previous PR passed locally and on ofborg (except aarch64-darwin which is still queued)

Change looks reasonable, if a bit janky

**Comment by RaitoBezarius** [comment]
Please do not merge yet
I'm planning to host the patch on webarchive and fetch it from there.

Le lun. 24 juil. 2023 à 02:39, Luna Nova ***@***.***> a
écrit :

**Comment by RaitoBezarius** [comment]
The patch has been moved into webarchive and fetched from there.

**Comment by LunNova** [comment]
@ofborg build nixosTests.boot.uefiCdrom

File: pkgs/development/compilers/edk2/default.nix
Comment (amjoseph-nixpkgs) [inline pkgs/development/compilers/edk2/default.nix:76]: Honestly even if it's big IMHO it needs to go in-tree, for lots of reasons.

If it absolutely can't go in-tree then it should go on tarballs.nixos.org.

`fetchpatch` FODs hash the normalized patch, not the thing that was fetched by `curl`... it's very much not an archival mechanism.  If `archive.org` decides to start mangling this patch, or adds banner ads to it or something silly like that, or just deletes it, we're going to have a nightmare of a time reproducing what went on here.

**Review by ghost — COMMENTED** [review]
Why can't we use the patch that (according to upstream's bugtracker) "was merged"?

There needs to be a lot more background in the commit message for something this wild and crazy.

File: pkgs/development/compilers/edk2/default.nix
Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/edk2/default.nix:76]: > Honestly if the patch is nixpkgs-specific then even if it's big it needs to go in-tree, for lots of reasons.

I just noticed that this PR is only for the `release-23.05` branch.

What's wrong with putting the diff in-tree in that case?  It won't become part of the long-term mainline nixpkgs, so it won't take up space for the rest of eternity.

Comment (RaitoBezarius) [reply pkgs/development/compilers/edk2/default.nix:76]: > It also doesn't look so good that the same person is (a) the nixpkgs UEFI expert (b) the author of this PR (c) the author of the patch (d) not hosting said patch themselves.

I would love if you would not do insinuations and be clear about what you would like out of it.

We have been going through some hoops to coordinate in some channels you were not, this PR may lack of context, but I also lack a lot of time, anyone is free to adopt this PR to expedite in a better fashion. I **really** do not appreciate going on me like this, especially with insinuations.

> Honestly if the patch is nixpkgs-specific then even if it's big it needs to go in-tree, for lots of reasons.

I would prefer this too and this was my first choice, after discussion, it was suggested to move it into webarchive like we do it for many other things in-tree, use grep to discover this.

> If it absolutely can't go in-tree then it should go on tarballs.nixos.org.
> fetchpatch FODs hash the normalized patch, not the thing that was fetched by curl... it's very much not an archival mechanism. If archive.org decides to start mangling this patch, or adds banner ads to it or something silly like that, or just deletes it, we're going to have a nightmare of a time reproducing what went on here.

Sure, but please take this matter in policy land, we are already using this and this is already a risk, plus, we are supposedly collecting via the find-tarballs script all the FODs and archiving them in tarballs.nixos.org, therefore, even this patch will end up there…

Comment (RaitoBezarius) [reply pkgs/development/compilers/edk2/default.nix:76]: Plus, you say in the review comment:

> Why can't we use the patch that (according to upstream's bugtracker) "was merged"?

As explained in the Nix file: https://github.com/NixOS/nixpkgs/pull/244727/files#diff-08eafc326951de5ee5ca50ad84a9b7b0668560086eb24c328dd270bea4db9bf4R57-R59

We cannot reuse as-is the patch, if you look at the delta, you will see that I only drop the `.gitmodules` part because fetchFromGitHub has no support for patching the git submodules section and applying this change while downloading the submodules, if this is not clear why this is non-trivial, I can make it clearer.

Comment (RaitoBezarius) [reply pkgs/development/compilers/edk2/default.nix:76]: And honestly if you want me to host the patch on my personal server, I can do it, I just don't think it's something we use to do and you will probably have to convince other people than me.

Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/edk2/default.nix:76]: > I would love if you would not do insinuations and be clear about what you would like out of it.

Sorry, **I did not mean to insinuate anything.**

Let me be totally explicit: *I* am not worried about this patch.  What I am worried about is how it looks to outsiders.  Also I am worried about other less-trustworthy people imitating this and pointing at this PR and saying "look it's okay @RaitoBezarius did it" and then I have to explain to them that I don't trust them as much as I trust you.


> As explained in the Nix file: https://github.com/NixOS/nixpkgs/pull/244727/files#diff-08eafc326951de5ee5ca50ad84a9b7b0668560086eb24c328dd270bea4db9bf4R57-R59

Is that the *only* obstacle to using upstream's patch verbatim?  I couldn't tell from the commit message (or this PR).  If that's the only obstacle I will figure out how to make this work.


> if you look at the delta, you will see that I only drop the `.gitmodules` part because fetchFromGitHub has no support for patching the git submodules section

(This is valuable information that should go in the commit message!)

Can't you just use `fetchpatch { excludes = ...;` to do that?  Then you could point the fetcher at upstream's bugzilla attachment and everything would be totally aboveboard.

If these were the only obstacles you encountered let me know and I can take over finishing this up; being release manager is a ton of work and this would be an easy way for me to lighten your burden.  After reading this PR I assumed the problem was some deep and fundamental UEFI thing rather than just "@(#*%@ nixpkgs `fetchpatch` and fetchers are a PITA".

Comment (RaitoBezarius) [reply pkgs/development/compilers/edk2/default.nix:76]: Aaaaaaah, I didn't know about `excludes`, give me a sec. :)

Comment (RaitoBezarius) [reply pkgs/development/compilers/edk2/default.nix]: Pushed a naive version with `excludes`, still fail with:

```
> applying patch /nix/store/ibzhl9qgac5n2gkd36zc28wgq33alj54-0021-Tweak-the-tools_def-to-support-cross-compiling.patch
       > patching file BaseTools/Conf/tools_def.template
       > Hunk #1 succeeded at 2294 (offset -5 lines).
       > Hunk #2 succeeded at 2326 (offset -5 lines).
       > applying patch /nix/store/lr9vyril0mwjs6w3ffyylfph6rgrnjlz-attachment.cgi?id=1330
       > patching file CryptoPkg/Library/OpensslLib/OpensslLib.inf
       > patching file CryptoPkg/Library/OpensslLib/OpensslLibAccel.inf
       > patching file CryptoPkg/Library/OpensslLib/OpensslLibCrypto.inf
       > patching file CryptoPkg/Library/OpensslLib/OpensslLibFull.inf
       > patching file CryptoPkg/Library/OpensslLib/OpensslLibFullAccel.inf
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/aes/aesni-mb-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/aes/aesni-sha1-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/aes/aesni-sha256-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/aes/aesni-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/aes/vpaes-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/modes/aesni-gcm-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/modes/ghash-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/sha/sha1-mb-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/sha/sha1-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/sha/sha256-mb-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/sha/sha256-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/sha/sha512-x86_64.S
       > patching file CryptoPkg/Library/OpensslLib/X64Gcc/crypto/x86_64cpuid.S
       > File CryptoPkg/Library/OpensslLib/openssl is not a regular file -- refusing to patch
       > 1 out of 1 hunk ignored -- saving rejects to file CryptoPkg/Library/OpensslLib/openssl.rej
```

Comment (RaitoBezarius) [reply pkgs/development/compilers/edk2/default.nix]: I will let you handle the magic :)

Comment (RaitoBezarius) [reply pkgs/development/compilers/edk2/default.nix]: And,

> Sorry, I did not mean to insinuate anything.

> Let me be totally explicit: I am not worried about this patch. What I am worried about is how it looks to outsiders. Also I am worried about other less-trustworthy people imitating this and pointing at this PR and saying "look it's okay @RaitoBezarius did it" and then I have to explain to them that I don't trust them as much as I trust you.

Thank you, you are totally right on that.

File: pkgs/development/compilers/edk2/default.nix
Comment (RaitoBezarius) [inline pkgs/development/compilers/edk2/default.nix:11]: needs to be dropped

File: pkgs/development/compilers/edk2/default.nix
Comment (RaitoBezarius) [reply pkgs/development/compilers/edk2/default.nix]: My intuition is that excludes requires `a/something` or `b/something`, I can do it via filterdiff, but I cannot reproduce it with `excludes` because I don't know what the normalized patch look like.

Comment (RaitoBezarius) [reply pkgs/development/compilers/edk2/default.nix]: Thanks to a @ncfavier tip, I discovered I just needed to bust the SRI hash to get the exclusion right because of the classical phenomena.

Comment (amjoseph-nixpkgs) [reply pkgs/development/compilers/edk2/default.nix]: > Thanks to a @ncfavier tip, I discovered I just needed to bust the SRI hash to get the exclusion right because of the classical phenomena.

Yeah that drives me nuts.  The fix is #188587 but it needs Nix>=2.12 and `impure-derivations`, which require `ca-derivations`, which is an irreversible upgrade to `/nix/store`, so it's going to be a while.

**Review by ghost — APPROVED** [review]
Very nice!

**Pull request merged**
